### PR TITLE
[RISCV][CostModel] Updates reduction and shuffle cost

### PR DESCRIPTION
**Pull request by arcbbb**
- Make `andi` cost 1 in SK_Broadcast
- Query the cost of VID_V, VRSUB_VX/VRSUB_VI which would scale with LMUL

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 1]
```diff
@@ -1358,19 +1363,53 @@ RISCVTTIImpl::getMinMaxReductionCost(Intrinsic::ID IID, VectorType *Ty,
...
-    // vcpop sequences, see vreduction-mask.ll.  umax, smin actually only
-    // cost 2, but we don't have enough info here so we slightly over cost.
-    return (LT.first - 1) + 3;
+  std::array<unsigned, 3> Opcodes;
+  if (Ty->getElementType()->isIntegerTy(1)) {
+    // vcpop sequences, see vreduction-mask.ll.
+    if ((IID == Intrinsic::umax) || (IID == Intrinsic::smin))
+      Opcodes = {RISCV::VMNAND_MM, RISCV::VCPOP_M, RISCV::SLT};
+    else
+      Opcodes = {RISCV::VCPOP_M, RISCV::SLT};
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1372]: The third element of `Opcodes` will be 0? The cost is not right then.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 2]
```diff
@@ -1392,20 +1431,50 @@ RISCVTTIImpl::getArithmeticReductionCost(unsigned Opcode, VectorType *Ty,
...
   std::pair<InstructionCost, MVT> LT = getTypeLegalizationCost(Ty);
-  if (Ty->getElementType()->isIntegerTy(1))
+  std::array<unsigned, 3> Opcodes;
+  if (Ty->getElementType()->isIntegerTy(1)) {
     // vcpop sequences, see vreduction-mask.ll
-    return (LT.first - 1) + (ISD == ISD::AND ? 3 : 2);
+    if (ISD == ISD::AND)
+      Opcodes = {RISCV::VMNAND_MM, RISCV::VCPOP_M, RISCV::SLT};
+    else
+      Opcodes = {RISCV::VCPOP_M, RISCV::SLT};
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1440]: Ditto.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 1 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1372]: ISD::SLT is not a vector instruction. Shouldn't it go through getCmpSelInstrCost?

Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: That 's indeed a bug. Thanks for pointing out! Fixed

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 2 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Fixed. Thanks!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 3]
```diff
@@ -1358,19 +1360,51 @@ RISCVTTIImpl::getMinMaxReductionCost(Intrinsic::ID IID, VectorType *Ty,
     return BaseT::getMinMaxReductionCost(IID, Ty, FMF, CostKind);
 
   std::pair<InstructionCost, MVT> LT = getTypeLegalizationCost(Ty);
-  if (Ty->getElementType()->isIntegerTy(1))
-    // vcpop sequences, see vreduction-mask.ll.  umax, smin actually only
-    // cost 2, but we don't have enough info here so we slightly over cost.
-    return (LT.first - 1) + 3;
+  SmallVector<unsigned, 3> Opcodes;
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1363]: Move this line below.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 4]
```diff
@@ -1392,20 +1426,58 @@ RISCVTTIImpl::getArithmeticReductionCost(unsigned Opcode, VectorType *Ty,
...
 
   std::pair<InstructionCost, MVT> LT = getTypeLegalizationCost(Ty);
-  if (Ty->getElementType()->isIntegerTy(1))
+  SmallVector<unsigned, 3> Opcodes;
+  Type *ElementTy = Ty->getElementType();
+  if (ElementTy->isIntegerTy(1)) {
     // vcpop sequences, see vreduction-mask.ll
-    return (LT.first - 1) + (ISD == ISD::AND ? 3 : 2);
+    if (ISD == ISD::AND) {
+      Opcodes = {RISCV::VMNAND_MM, RISCV::VCPOP_M};
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1434]: Same problem caused by uninitialized elements.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 5]
```diff
@@ -1392,20 +1426,58 @@ RISCVTTIImpl::getArithmeticReductionCost(unsigned Opcode, VectorType *Ty,
...
+      return (LT.first - 1) +
+             getRISCVInstructionCost(Opcodes, LT.second, CostKind) +
+             getCmpSelInstrCost(Instruction::Select, ElementTy, ElementTy,
+                                CmpInst::ICMP_EQ, CostKind);
+    } else {
+      Opcodes = {RISCV::VCPOP_M};
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1440]: Ditto.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 4 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1434]: You've made a good observation about the capacity and size here.
For clarity, in the context of SmallVector the capacity of Opcodes is 3, and the current size is 2 here.
ArrayRef is constructed from the size of the SmallVector, so I think the extra capacity will not be used.
I was based on the logic in
```
    /// Construct an ArrayRef from a SmallVector. This is templated in order to
    /// avoid instantiating SmallVectorTemplateCommon<T> whenever we
    /// copy-construct an ArrayRef.
    template<typename U>
    /*implicit*/ ArrayRef(const SmallVectorTemplateCommon<T, U> &Vec)
      : Data(Vec.data()), Length(Vec.size()) {
    }
```

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 3 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Fixed!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 4 above)
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1434]: Oh I misread it, I thought it was std::array.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 6]
```diff
@@ -84,8 +84,14 @@ RISCVTTIImpl::getRISCVInstructionCost(ArrayRef<unsigned> OpCodes, MVT VT,
       Cost += VL;
       break;
     }
+    case RISCV::VMV_X_S:
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:87]: These four cases can be merged.

**Review by wangpc-pp — APPROVED** [review]
LGTM with nits.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 7]
```diff
@@ -1392,20 +1426,58 @@ RISCVTTIImpl::getArithmeticReductionCost(unsigned Opcode, VectorType *Ty,
...
+  SmallVector<unsigned, 3> Opcodes;
+  Type *ElementTy = Ty->getElementType();
+  if (ElementTy->isIntegerTy(1)) {
     // vcpop sequences, see vreduction-mask.ll
-    return (LT.first - 1) + (ISD == ISD::AND ? 3 : 2);
+    if (ISD == ISD::AND) {
+      Opcodes = {RISCV::VMNAND_MM, RISCV::VCPOP_M};
+      return (LT.first - 1) +
+             getRISCVInstructionCost(Opcodes, LT.second, CostKind) +
+             getCmpSelInstrCost(Instruction::Select, ElementTy, ElementTy,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1437]: Shouldn't this be Instruction::ICmp?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 8]
```diff
@@ -1392,20 +1426,58 @@ RISCVTTIImpl::getArithmeticReductionCost(unsigned Opcode, VectorType *Ty,
...
+    Opcodes = {RISCV::VMV_S_X, RISCV::VREDXOR_VS, RISCV::VMV_X_S};
+    break;
+  case ISD::AND:
+    SplitOp = RISCV::VAND_VV;
+    Opcodes = {RISCV::VMV_S_X, RISCV::VREDAND_VS, RISCV::VMV_X_S};
+    break;
+  case ISD::FADD:
+    SplitOp = RISCV::VFADD_VV;
+    if (TTI::requiresOrderedReduction(FMF))
+      Opcodes = {RISCV::VFMV_S_F, RISCV::VFREDOSUM_VS, RISCV::VFMV_F_S};
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1470]: Ordered reductions don't split the same way. They chain a series of VFREDOSUMs.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 7 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Fixed. Thanks!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 8 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Thanks for pointing out this! Fixed.

**Comment by preames** [comment]
I split one piece of this off in https://github.com/llvm/llvm-project/pull/77963.

For this to make rapid progress, I recommend splitting it into individual pieces to the maximum extent possible.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 9]
```diff
@@ -84,8 +84,12 @@ RISCVTTIImpl::getRISCVInstructionCost(ArrayRef<unsigned> OpCodes, MVT VT,
       Cost += VL;
       break;
     }
+    case RISCV::VMV_X_S:
+    case RISCV::VFMV_F_S:
     case RISCV::VMV_S_X:
-      // FIXME: VMV_S_X doesn't use LMUL, the cost should be 1
+    case RISCV::VFMV_S_F:
+      Cost += 1;
+      break;
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:92]: Can we split this off into a separate PR? I think it would be useful to be able to see if this has any effect on the current cost model tests.

Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:92]: Sure, I split it in #78739. will update this PR after that.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 10]
```diff
@@ -523,9 +523,9 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
     if (LT.second.isFixedLengthVector())
       // vrsub.vi has a 5 bit immediate field, otherwise an li suffices
       LenCost = isInt<5>(LT.second.getVectorNumElements() - 1) ? 0 : 1;
-    // FIXME: replace the constant `2` below with cost of {VID_V,VRSUB_VX}
-    InstructionCost GatherCost =
-        2 + getRISCVInstructionCost(RISCV::VRGATHER_VV, LT.second, CostKind);
+    InstructionCost GatherCost = getRISCVInstructionCost(
+        {RISCV::VID_V, RISCV::VRSUB_VX, RISCV::VRGATHER_VV}, LT.second,
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:527]: This bit isn't right.  In particular, it ignore the code two lines above which talks about different cases for costing the index sequence.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 11]
```diff
@@ -1358,19 +1358,51 @@ RISCVTTIImpl::getMinMaxReductionCost(Intrinsic::ID IID, VectorType *Ty,
     return BaseT::getMinMaxReductionCost(IID, Ty, FMF, CostKind);
 
   std::pair<InstructionCost, MVT> LT = getTypeLegalizationCost(Ty);
-  if (Ty->getElementType()->isIntegerTy(1))
-    // vcpop sequences, see vreduction-mask.ll.  umax, smin actually only
-    // cost 2, but we don't have enough info here so we slightly over cost.
-    return (LT.first - 1) + 3;
+  if (Ty->getElementType()->isIntegerTy(1)) {
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1361]: If you separate this as it's own patch, should be easy to review.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 12]
```diff
@@ -1358,19 +1358,51 @@ RISCVTTIImpl::getMinMaxReductionCost(Intrinsic::ID IID, VectorType *Ty,
...
+      return getArithmeticReductionCost(Instruction::And, Ty, FMF, CostKind);
+  }
 
   // IR Reduction is composed by two vmv and one rvv reduction instruction.
-  InstructionCost BaseCost = 2;
-
-  if (CostKind == TTI::TCK_CodeSize)
-    return (LT.first - 1) + BaseCost;
-
-  unsigned VL = getEstimatedVLFor(Ty);
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1372]: The net effect of this change appears to be to increase BaseCost by 1, make the reduction linear in lmul, and remove the log2(VL) term.  

This bit (maybe with the other reduction change) needs to be it's own review so we can clearly see the effect.  This one (in particular), you should also try running through some reasonable large codebase (llvm test-suite, spec, whatever) and checking for unexpected interactions.

Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1372]: OK! I split this off into #79103.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 11 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1361]: It is split off in https://github.com/llvm/llvm-project/pull/79401

**Comment by preames** [comment]
After https://github.com/llvm/llvm-project/pull/79402 lands, I think all the pieces of this have gone in right?  If not, can you rebase so what's left is visible?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 10 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: I made a change to query the cost of VRSUB_VX/VRSUB_VI separately. Not sure this is what you expected.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 13]
```diff
@@ -482,11 +481,10 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
       //   vmsne.vi  v0, v8, 0
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:481]: Unrelated codegen opportunity

 vmv.x.s a0, v8
 andi    a0, a0, 1
 vmv.v.x v8, a0

Could be a vrgather.vi

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 14]
```diff
@@ -482,11 +481,10 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
 
       return LT.first *
-             (TLI->getLMULCost(LT.second) + // FIXME: this should be 1 for andi
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:485]: If the index is non-zero, then there's a vslide needed in the current sequence which does scale with LMUL.

So, this change is wrong for non-zero index.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 15]
```diff
@@ -528,12 +526,19 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
     InstructionCost LenCost = 3;
-    if (LT.second.isFixedLengthVector())
+    InstructionCost GatherCost;
+    if (LT.second.isFixedLengthVector()) {
       // vrsub.vi has a 5 bit immediate field, otherwise an li suffices
       LenCost = isInt<5>(LT.second.getVectorNumElements() - 1) ? 0 : 1;
-    // FIXME: replace the constant `2` below with cost of {VID_V,VRSUB_VX}
-    InstructionCost GatherCost =
-        2 + getRISCVInstructionCost(RISCV::VRGATHER_VV, LT.second, CostKind);
+      GatherCost = getRISCVInstructionCost(
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:533]: Please common the tail here, and undo the change which is restructuring the existing LenCost computation since that appears to be non-functional in your diff.

**Review by preames — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 14 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:485]: For SK_Broadcast, my understanding is that the index should always be the first element of the source vector. 
I was based on this
```
      if (Shuffle->isZeroEltSplat())
        return TargetTTI->getShuffleCost(TTI::SK_Broadcast, VecTy,
                                         Shuffle->getShuffleMask(), CostKind, 0,
                                         nullptr, Operands);
```
Could we clarify this?"

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 15 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Fixed

**Comment by arcbbb** [comment]
> Given you've now reduced the scope of this, you need to update the review description so that it matches the current patch.

Updated, thanks!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 14 above)
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:485]: It looks like you're correct here.  There are other cases which create SK_Broadcast, but from what I can tell, all are consistent with the splat index always being zero.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 16]
```diff
@@ -531,9 +529,12 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
     if (LT.second.isFixedLengthVector())
       // vrsub.vi has a 5 bit immediate field, otherwise an li suffices
       LenCost = isInt<5>(LT.second.getVectorNumElements() - 1) ? 0 : 1;
-    // FIXME: replace the constant `2` below with cost of {VID_V,VRSUB_VX}
+    unsigned Opcodes[] = {RISCV::VID_V, RISCV::VRSUB_VX, RISCV : VRGATHER_VV};
```
[the remark this answers is not in the document]
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Maybe I overlooked something, here are the three cost sequences I am considering:
1. scalable vector
   LenCost:3 from { csrr a0, vlenb, srli a0, a0, 3, addi a0, a0, -1 }
   GatherCost from {vid.v, vrsub.vx, vrgather.vv}
2. fixed-length vector, imm > 5bit
    LenCost:1 from { li }
    GatherCost from {vid.v, vrsub.vx, vrgather.vv}
3. fixed-length vector, imm <= 5 bit
    LenCost:0 from { }
    GatherCost from {vid.v, vrsub.vi, vrgather.vv}

Could you help me correct this?

[the remark this answers is not in the document]
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: @preames I can't see where LenCost is changed in this branch. I think you might be looking at an earlier version of the diff before this force push? https://github.com/llvm/llvm-project/commit/62be3d6f1b95b0554a1ae0b248f3c5065b9bbba2

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 14 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:485]: To chime in here, it's defined in TargetTransformInfo.h as being a broadcast of element zero:

```c++
enum ShuffleKind {
    SK_Broadcast,        ///< Broadcast element 0 to all other elements.
```

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 17]
```diff
@@ -531,9 +529,12 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
     if (LT.second.isFixedLengthVector())
       // vrsub.vi has a 5 bit immediate field, otherwise an li suffices
       LenCost = isInt<5>(LT.second.getVectorNumElements() - 1) ? 0 : 1;
-    // FIXME: replace the constant `2` below with cost of {VID_V,VRSUB_VX}
+    unsigned Opcodes[] = {RISCV::VID_V, RISCV::VRSUB_VX, RISCV::VRGATHER_VV};
+    if (LT.second.isFixedLengthVector() &&
+        isInt<5>(LT.second.getVectorNumElements() - 1))
+      Opcodes[1] = RISCV::VRSUB_VI;
```
[the remark this answers is not in the document]
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:535]: Yes, the cost for both is the same.

**Comment by arcbbb** [comment]
@preames ping

**Review by lukel97 — APPROVED** [review]
LGTM

**Comment by arcbbb** [comment]
I am going to merge this if no further inputs. Thanks!
